01-resource-referral/langgraph-workflow.py

```diff
@@ -0,0 +1,242 @@
+        graph.add_node("call_tool_spreadsheet", self.call_spreadsheet_tool)
+        graph.add_edge("run_llms", "llm_spreadsheet_query")
+        graph.add_edge("llm_spreadsheet_query", "call_tool_spreadsheet")
+
+        graph.add_edge("call_tool_211_api", "decision_node")
+        graph.add_edge("call_tool_spreadsheet", "decision_node")
+        self._draw_graph(graph)
+        return graph
+
+    def _draw_graph(self, graph: StateGraph):
+        g = graphviz.Digraph(filename="workflow.dot")
+        g.node(END, shape="box")
+        g.node("decision_node", style="filled", color="gray")
+        for name, _ in graph.nodes.items():
+            g.node(name)
+        for n1, n2 in graph.edges:
+            g.edge(n1, n2)
+        # debug_here(locals())
+        for n, branches in graph.branches.items():
+            for bb in branches:
+                g.edge(n, bb.condition.__name__, style="dotted")
+        g.node("decide_next_node", shape="diamond")
+        # Add edges based on decide_next_node() return value
+        g.edge("decide_next_node", END, style="dotted")
+        g.edge("decide_next_node", "merge_tool_results", style="dotted")
+        g.edge("decide_next_node", "run_llms", style="dotted")
+        # Paste printout into https://dreampuf.github.io/GraphvizOnline to visualize graph
+        print(g.source)
+
+    # Determines next node to call
+    def decide_next_node(self, state):
+        print("\nNEXT_EDGE") # , json.dumps(state, indent=2))
+
+        if state["final_answer"]:
+            return END
+
+        if self._got_responses_from_all_tools(state):
+            return "merge_tool_results"
+
+        return "run_llms"
+
+    def check_for_final_answer(self, state):
+        print("\nHAS_FINAL_ANSWER node")  # , json.dumps(state, indent=2))
+
+        last_message = state["messages"][-1]
+        if "user_query" in last_message:
+            return None
+
+        if state["final_answer"]:
+            return None
+
+        if self._got_responses_from_all_tools(state):
+            # Return a list for "messages" because it will get added to the existing list
+            return {"messages": [{"log": "Got both responses"}]}
+
+        # This doesn't occur probably b/c LangGraph waits for all incoming edges to be complete
+        print("\nHAS_FINAL_ANSWER node: Waiting for more responses")
+
+    def _got_responses_from_all_tools(self, state):
+        expected_tools = [ "spreadsheet", "211_api" ]
+        return all(key in state["tool_responses"] for key in expected_tools)
+
+    def run_llms(self, state):
+        print("\nRUN_LLMS node", state["messages"][-1])
+        return None
+
+    def llm_211_api(self, state):
+        # find the last message that mentions the "user_query"
+        user_message = next(m for m in reversed(state["messages"]) if "user_query" in m)
+        print("\nLLM_211_API node", user_message)
+
+        # Human-in-the-loop example:
```

Comment:
did we want to remove this?

Reply:
I don't mind `Human-in-the-loop example` being there. What are your thoughts?

Reply:
I'm open to having it there, but maybe we want to handle the "n" option differently as opposed to throwing a ValueError. I'm currently going through the comments and seeing where we can do some clean up

Reply:
Gotcha, how about a call to `sys.exit()` instead of the `ValueError`?